libkbfs/block_journal.go

```diff
@@ -978,179 +945,85 @@ func (j *blockJournal) ignoreBlocksAndMDRevMarkers(ctx context.Context,
 		idsToIgnore[id] = true
 	}
 
-	err := j.ignoreBlocksAndMDRevMarkersInJournal(ctx, idsToIgnore, rev, j.j)
-	if err != nil {
-		return err
-	}
-
-	if j.saveUntilMDFlush == nil {
-		return nil
-	}
-
-	return j.ignoreBlocksAndMDRevMarkersInJournal(
-		ctx, idsToIgnore, rev, *j.saveUntilMDFlush)
-}
-
-func (j *blockJournal) saveBlocksUntilNextMDFlush() error {
-	if j.saveUntilMDFlush != nil {
-		return nil
-	}
-
-	// Copy the current journal entries into a new journal.  After the
-	// next MD flush, we can use the saved journal to delete the block
-	// data for all the entries in the saved journal.
-	first, err := j.j.readEarliestOrdinal()
-	if ioutil.IsNotExist(err) {
-		return nil
-	} else if err != nil {
-		return err
-	}
-
-	last, err := j.j.readLatestOrdinal()
-	if ioutil.IsNotExist(err) {
-		return nil
-	} else if err != nil {
-		return err
-	}
-
-	savedJournalDir := savedBlockJournalDir(j.dir)
-	sj := makeDiskJournal(
-		j.codec, savedJournalDir, reflect.TypeOf(blockJournalEntry{}))
-	savedJournal := &sj
-
-	for i := first; i <= last; i++ {
-		e, err := j.readJournalEntry(i)
-		if err != nil {
-			return err
-		}
-
-		savedJournal.appendJournalEntry(nil, e)
-	}
-
-	j.saveUntilMDFlush = savedJournal
-	return nil
+	return j.ignoreBlocksAndMDRevMarkersInJournal(ctx, idsToIgnore, rev, j.j)
 }
 
-// onMDFlush removes at most `maxToRemove` blocks from the
-// `saveUntilMDFlush` journal if one exists.  If `lastToRemove` is
-// zero, it flushes the complete `saveUntilMDFlush` if it has fewer
-// than `maxToRemove` entries; if it doesn't flush the entire journal,
-// it returns the ordinal of the current last entry.  If
-// `lastToRemove` is non-zero, it only flushes up to the minimum of
-// `lastToRemove` and the earliest entry + `maxToRemove`; if this
-// flushes the entire saved journal or reaches a non-ignored marker
-// for MD revision `flushedMDRev`, it returns 0, otherwise it returns
-// `lastToRemove`.  It's intended that the caller should call this
-// function repeatedly until it returns 0, releasing any locks in
-// between calls so it doesn't block other operations for too long.
-func (j *blockJournal) onMDFlush(ctx context.Context,
-	maxToRemove uint64, flushedMDRev MetadataRevision,
-	lastToRemove journalOrdinal) (
-	nextLastToRemove journalOrdinal,
-	removedBytes, removedFiles int64, err error) {
-	if j.saveUntilMDFlush == nil {
-		return 0, 0, 0, nil
-	}
-
-	if maxToRemove == 0 {
-		return 0, 0, 0, errors.New("maxToRemove must be non-zero")
-	}
-
-	// Delete the block data for anything in the saved journal.
-	first, err := j.saveUntilMDFlush.readEarliestOrdinal()
+// getDeferredRange gets the earliest and latest revision of the
+// deferred GC journal.  If the returned length is 0, there's no need
+// for further GC.  Should be called under lock.
```

Comment:
the block journal doesn't know about locking, so I don't know if that last sentence should be there. There's a type-level comment about this type not being goroutine safe, so maybe only `onMDFlush` needs the special comment.